src/state.py

```diff
@@ -0,0 +1,155 @@
+                    (
+                        "Each Hostname segment must be less than 64 in length, and consist of "
+                        "alphanumeric and hyphen"
+                    )
+                )
+
+        return value
+
+    @pydantic.field_validator("path")
+    @classmethod
+    def validate_path(cls, value: str) -> str:
+        """Validate the path.
+
+        Args:
+            value: The value to validate.
+
+        Raises:
+            ValueError: The validation failed.
+
+        Returns:
+            The value after validation.
+        """
+        # This are the valid characters for path in addition to `/`:
+        # a-z A-Z 0-9 . - _ ~ ! $ & ' ( ) * + , ; = : @
+        # https://datatracker.ietf.org/doc/html/rfc3986#section-3.3
+        valid_path = re.compile(r"[/\w.\-~!$&'()*+,;=:@]+", re.IGNORECASE)
+        if valid_path.fullmatch(value) is None:
+            raise ValueError("Path contains non-allowed character")
+        return value
+
+    @classmethod
+    def from_charm(cls, charm: ops.CharmBase) -> "Configuration":
+        """Initialize object from the charm.
+
+        Args:
+            charm: The charm containing the configuration.
+
+        Raises:
+            ConfigurationError: Error with the charm configuration.
+
+        Returns:
+            The object.
+        """
+        hostname = typing.cast(str, charm.config.get(HOSTNAME_CONFIG_NAME, "")).strip()
+        path = typing.cast(str, charm.config.get(PATH_CONFIG_NAME, "")).strip()
+        protocol = typing.cast(str, charm.config.get(PROTOCOL_CONFIG_NAME, "")).lower().strip()
+        backends_str = typing.cast(str, charm.config.get(BACKENDS_CONFIG_NAME, "")).strip()
+        if not backends_str:
+            raise ConfigurationError("Empty backends configuration found")
+
+        backends = tuple(ip.strip() for ip in backends_str.split(","))
+        try:
+            return cls(
+                hostname=hostname,
+                path=path,
+                # Pydantic allows converting str to IPvAnyAddress.
+                backends=backends,  # type: ignore
+                # Pydantic allows converting str to a string enum.
+                protocol=protocol,  # type: ignore
+            )
+        except pydantic.ValidationError as err:
+            err_msg = [
+                f'{error["loc"][0]} = {error["input"]}: {error["msg"]}' for error in err.errors()
+            ]
+            logger.error("Found config error: %s", err_msg)
+            raise ConfigurationError(f"Config error: {err_msg}") from err
+
+    def to_integration_data(self) -> dict[str, str]:
+        """Convert to format supported by integration.
+
+        Juju integration only supports data of dict[str, str] type.
+        This method ensures the the values in the dict are all str type.
+
+        Returns:
+            The data in the format accepted by integrations.
+        """
+        data = json.loads(self.model_dump_json())
```

Comment:
What happens in the case of json.loads raising an exception?

Reply:
I have added guards against json related errors.
For this case, I think the pydantic `model_dump_json` should be producing valid json. I have added exception catching for `PydanticSerializationError` to guard against `model_dump_json` errors.